dashboard/app/controllers/regional_partners_controller.rb

```diff
@@ -109,6 +114,35 @@ def remove_mapping
     redirect_to @regional_partner
   end
 
+  # POST /regional_partners/:id/replace_mappings
+  # Replace mappings for partner with given id with contents
+  # of CSV file passed in regions parameter
+  # CSV file must have header 'Region' and all regions must be
+  # valid (zip code or state), or no replacement will be done
+  def replace_mappings
+    regions_file = params[:regions]
+    csv = validate_mappings_csv(regions_file)
+    unless csv
+      redirect_to @regional_partner
+      return
+    end
+
+    mappings = []
+    errors = []
+    parse_mappings_from_csv(csv, mappings, errors)
+
+    if !errors.empty?
+      flash[:upload_error] = parse_upload_errors(errors)
+      redirect_to @regional_partner
+    else
+      # use destroy_all to ensure old mappings are soft deleted
+      @regional_partner.mappings.destroy_all
+      @regional_partner.mappings = mappings
```

Comment:
It's driving me crazy that I don't understand exactly where the new mappings are saved. I think this line is where the `save` operation for every built mapping occurs, but I'm having trouble finding documentation for this behavior.  There are some notes in [`ActiveRecord::Associations::ClassMethods`](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html) under [Unsaved objects and associations > Collections](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#module-ActiveRecord::Associations::ClassMethods-label-Collections) that sort of document this behavior, but all the examples in the documentation of [Active Record Autosave Associations](https://api.rubyonrails.org/classes/ActiveRecord/AutosaveAssociation.html) depend on `build` implicitly adding to the parent object's collection and then calling `save` on the parent object.

```rb
post = Post.new(title: 'ruby rocks')
post.comments.build(body: 'hello world')
post.save # => saves both post and comment
```

And after looking at that example, I think `parse_mappings_from_csv` _does_ actually modify `@regional_partner.mappings` even though it doesn't _save_ the new mappings, and in turn I'm pleasantly surprised that `destroy_all` doesn't set a delete marker on the new, unsaved mappings.

All that to say, my understanding of what's going on here is less precise than I'd like, and I'm wondering if you found a better example someplace that I could read?

Reply:
Good question. As for where it is actually saved, the `@regional_partner.mappings = mappings` does that ([reference](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-many-collection-objects))
Your question answers some weird behavior I saw while trying to make this work--I had found that calling clear worked to remove the old mappings (but didn't soft delete), but running delete_all or trying to individually delete the mappings in @regional_partner would delete the new ones too. Now I'm not sure why destroy_all doesn't do that--perhaps because it invokes callbacks whereas delete doesn't?

Reply:
Hmm, puzzling.  Thank you for the documentation reference, that clears things up.  I think your approach is fine, especially with the good unit tests in place.

It might be a good idea to wrap these two lines in a transaction.  I know you've checked all the validations in advance, but just in case something really unexpected goes wrong between `destroy_all` and `collection=` (like a dropped DB connection) it'd be one more thing keeping us from landing in a bad state.

Reply:
added transaction